doc/functions/fileset.section.md

```diff
@@ -0,0 +1,117 @@
+# File sets {#sec-fileset}
+
+File sets efficiently represent a set of local files. File sets easily be created, combined and modified for complex behavior. All file set functionality is exposed under `lib.fileset`.
+
+The most basic way to create file sets is by passing a path to `lib.singleton`. The resulting file set depends on the path:
+- If the path points to a file, the result is a file set only consisting of that single file
+- If the path points to a directory, all files in that directory will be in the resulting file set
+
+This matches the list of files that would be imported when using string interpolation on the path, e.g. `"${./Makefile}"` or `"${./src}"`.
+
+Another way to create file sets is by combining existing file sets together, e.g. using `lib.fileset.unions`:
+
+```nix
+with lib.fileset;
+unions [
+  (singleton ./Makefile)
+  (singleton ./src)
+]
+```
+
+For convenience, all fileset operations implicitly coerce path arguments to file sets using `lib.singleton`, so the above is the same as
+```nix
+with lib.fileset;
+unions [
+  ./Makefile
+  ./src
+]
+```
+
+Another way to create file sets is by transforming existing file sets, e.g. using `lib.fileset.suffixFileFilter`:
+```
+with lib.fileset;
+union [
+  ./Makefile
+  (suffixFileFilter ["c" "h"] ./src)
+]
+```
+
+A fileset can be used in a derivation build by passing it as a `src` to any `stdenv.mkDerivation`-based builder. It is then necessary to also set `sourceRoot` to the path of the builder entry point.
+```nix
+stdenv.mkDerivation {
+  src =
+    with lib.fileset;
+    unions [
+      ./Makefile
+      (suffixFileFilter ["c" "h"] ./src)
+    ];
+  sourceRoot = ./.;
+  postUnpack = ''
+    cat ./Makefile
+    ls ./src
+  '';
+}
+```
+
+By setting `sourceRoot` to `./.`, the builder can access the files at paths relative to directory of the file that defines it.
+
+Note that only exactly the files in the file set will be imported into the store and influence the derivation hash.
+
+File sets are not restricted to just files in the current directory though, any file can be included. If you're in a project subdirectory this means that you can use also include files from the project root directory:
+```nix
+stdenv.mkDerivation {
+  src =
+    with lib.fileset;
+    unions [
+      ./Makefile
+      (suffixFileFilter ["c" "h"] ./src)
+      ../README.md
+    ];
+  sourceRoot = ./.;
+  postUnpack = ''
+    cat ./Makefile
+    ls ./src
+    cat ../README.md
+  '';
+}
+```
+
+Warning: Any intermediate directory names will influence the store hash.
+
+Here's a more extensive example showcasing some file set functions:
+
+```nix
+stdenv.mkDerivation {
+
+  src =
+    with lib.fileset;
+
+    # Intersection between two sets of files
+    intersect
+
+      # All files tracked by the git root in the parent
+      (gitTracked ../.)
+
+      # Remove all files in "manual" directories
+      (lib.flip difference (directoryFilter "manual" ./.)
```

Comment:
Should we have a `difference` with a flipped order?  The thing to subtract is often smaller than the thing you want to subtract from (as here), so that would be convenient. But what should the name then be?